### ISSUE TITLE
Google Civic API integration

### ISSUE THREAD

**Issue by spencersnygg** [spencersnygg/VoteWise#13]
I would like integration with Google's civic API so that we can have the users type in their address and have our database populated with the info on what districts they live in.  I mention this need also in the Create Voter User page, but we need it in other places too and it's a huge job.

**Comment by rms2219** [comment]
Spencer, what's the need to store the info from the Google Civic API? Is
there a reason we can't pull it in on-demand from the user info (zip code)?

**Comment by msalas74** [comment]
We should not store any data since we are utilizing the API.  This can be accomplished through the frontend vs the backend.  That way the backend resource will be limited.  It should really be setup for a frontend process and display the information to each specific user's geographic location.  That data shouldnt really reside in our database.  It would just be a copy of Google's database.


On Aug 9, 2015, at 11:13 AM, rms2219 wrote:

> Spencer, what's the need to store the info from the Google Civic API? Is
> there a reason we can't pull it in on-demand from the user info (zip code)?
> 
> On Sun, Aug 9, 2015, 2:06 PM Spencer Snygg <notifications@github.com> wrote:
> 
> > I would like integration with Google's civic API so that we can have the
> > users type in their address and have our database populated with the info
> > on what districts they live in. I mention this need also in the Create
> > Voter User page, but we need it in other places too and it's a huge job.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/spencersnygg/VoteWise/issues/13>.
> >
> —
> Reply to this email directly or view it on GitHub.
> 

Sincerely,

Matias Salas

HM: 510-894-1697
MOBILE: 510-688-9996

**Comment by spencersnygg** [comment]
Unfortunately, the voter districts are not necessarily divided by zip
code.  Also unfortunately, Google's Civic API seems to be the only gun in
town with the ability to figure out all those subdivisions in one step. 
Every state has its own way of doing things, every county has it's own way
of doing things.  Every city has its own way of doing things.  Google put
it all in one box for us to use.

Google Civic API should be able to provide which elections the voter is
eligible to vote for and who the candidates are (or at least who the
incumbents are).

-Spencer

**Comment by spencersnygg** [comment]
We need to store what voter districts the user resides in so we can match
them up with politicians in those districts.

**Comment by spencersnygg** [comment]
Yeah, we don't want to copy google's database.  we just want to make use of it and store the output for that particular user.

**Comment by msalas74** [comment]
I believe that we can do this without storing the data.  If we store every users info to our servers database, it might have some financial repercussion since this action will potentially require a large bandwidth.  Most host providers will charge for that service.

On Aug 9, 2015, at 11:28 AM, Spencer Snygg wrote:

> We need to store what voter districts the user resides in so we can match
> them up with politicians in those districts.
> 
> > We should not store any data since we are utilizing the API. This can be
> > accomplished through the frontend vs the backend. That way the backend
> > resource will be limited. It should really be setup for a frontend
> > process and display the information to each specific user's geographic
> > location. That data shouldnt really reside in our database. It would
> > just be a copy of Google's database.
> >
> >
> > On Aug 9, 2015, at 11:13 AM, rms2219 wrote:
> >
> >> Spencer, what's the need to store the info from the Google Civic API? Is
> >> there a reason we can't pull it in on-demand from the user info (zip
> >> code)?
> >>
> >> On Sun, Aug 9, 2015, 2:06 PM Spencer Snygg <notifications@github.com>
> >> wrote:
> >>
> >> > I would like integration with Google's civic API so that we can have
> >> the
> >> > users type in their address and have our database populated with the
> >> info
> >> > on what districts they live in. I mention this need also in the Create
> >> > Voter User page, but we need it in other places too and it's a huge
> >> job.
> >> >
> >> > —
> >> > Reply to this email directly or view it on GitHub
> >> > <https://github.com/spencersnygg/VoteWise/issues/13>.
> >> >
> >> —
> >> Reply to this email directly or view it on GitHub.
> >>
> >
> > Sincerely,
> >
> > Matias Salas
> >
> > HM: 510-894-1697
> > MOBILE: 510-688-9996
> >
> >
> >
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/spencersnygg/VoteWise/issues/13#issuecomment-129223103
> 
> 
> —
> Reply to this email directly or view it on GitHub.
> 

Sincerely,

Matias Salas

HM: 510-894-1697
MOBILE: 510-688-9996

**Comment by msalas74** [comment]
we can just pull the data, do something with it and present it.  this can be done dynamically and without saving the data.


On Aug 9, 2015, at 11:28 AM, Spencer Snygg wrote:

> We need to store what voter districts the user resides in so we can match
> them up with politicians in those districts.
> 
> > We should not store any data since we are utilizing the API. This can be
> > accomplished through the frontend vs the backend. That way the backend
> > resource will be limited. It should really be setup for a frontend
> > process and display the information to each specific user's geographic
> > location. That data shouldnt really reside in our database. It would
> > just be a copy of Google's database.
> >
> >
> > On Aug 9, 2015, at 11:13 AM, rms2219 wrote:
> >
> >> Spencer, what's the need to store the info from the Google Civic API? Is
> >> there a reason we can't pull it in on-demand from the user info (zip
> >> code)?
> >>
> >> On Sun, Aug 9, 2015, 2:06 PM Spencer Snygg <notifications@github.com>
> >> wrote:
> >>
> >> > I would like integration with Google's civic API so that we can have
> >> the
> >> > users type in their address and have our database populated with the
> >> info
> >> > on what districts they live in. I mention this need also in the Create
> >> > Voter User page, but we need it in other places too and it's a huge
> >> job.
> >> >
> >> > —
> >> > Reply to this email directly or view it on GitHub
> >> > <https://github.com/spencersnygg/VoteWise/issues/13>.
> >> >
> >> —
> >> Reply to this email directly or view it on GitHub.
> >>
> >
> > Sincerely,
> >
> > Matias Salas
> >
> > HM: 510-894-1697
> > MOBILE: 510-688-9996
> >
> >
> >
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/spencersnygg/VoteWise/issues/13#issuecomment-129223103
> 
> 
> —
> Reply to this email directly or view it on GitHub.
> 

Sincerely,

Matias Salas

HM: 510-894-1697
MOBILE: 510-688-9996

**Comment by spencersnygg** [comment]
Google's CIVIC API would be one stop shopping to get the voter's district divisions. (Federal congressional district, state congressional district, county, city, school district....)  Whether we automatically store this information, or have the user type it in, it is critical information.  We need to store that information in order for the site to work.

**Comment by spencersnygg** [comment]
We don't need to store a lot of other info that google provides like pictures of the candidates and other stuff like that.  Just the fields I mentioned.

**Comment by spencersnygg** [comment]
By the nature of answering tons of questions, and getting a match-up. we will have tons of text info back and forth stored on the server and displayed.  That is a cost of the site.  Microsoft has agreed to foot the bill for 3 years if we have a small amount of use or one year if we have a huge amount of use.

**Comment by rms2219** [comment]
What Matias said is what I was getting at. There's no reason to store the
result of the Civic API request when we can just query it on demand using
the data supplied by our users. We'll store whatever user information is
necessary to make that query. Another reason you may not want to store this
data is because we are then responsible for invalidating it once something
changes at the source. Now, we could certainly cache the result for some
period so that we can cut down on API requests.

**Comment by spencersnygg** [comment]
You are saying we shouldn't store what district the users live in?  It's not that much info.  Without storing it, we would have to re-ask them everytime they logged in.  And we would have no way of knowing if they were intentionally screwing up polling data.  Please explain to me the benefit of not storing this information.

**Comment by spencersnygg** [comment]
Personally, I know my ward number, but I don't know my state or federal congressional district numbers, so if the site didn't store that info, I would have to look it up every time and that would be a big pain in the butt.

**Comment by msalas74** [comment]
We would just need the users address to get all that information from the API's database.  I'm assuming that the users address will be collected when they sign up for a user account.  Since we would then have that data in our database, it's all we need to obtain the information regarding the users district.  Google maintains that database and if we decide to keep copies (even if just a few) we will need to code in for obtaining the latest information.  Since we should do this anyways during the users login process (and maybe even incorporate a schedule for retrieving a new copy for the app for those that stay logged in for a duration)  I just dont see the need to add the data to our database since Google is already doing the database management.

Here's Goggles description on how the API works - 
"Our data is based on the political geography of a citizen’s address. This address indicates where a citizen is eligible to vote and who represents them. There are many U.S. elections throughout the year, and both election information and political geography can shift with time. Google assigns every election available in the API an election ID, and the information associated with that ID is intended to be accurate for that election only."


On Aug 9, 2015, at 11:39 AM, Spencer Snygg wrote:

> Google's CIVIC API would be one stop shopping to get the voter's district divisions. (Federal congressional district, state congressional district, county, city, school district....) Whether we automatically store this information, or have the user type it in, it is critical information. We need to store that information in order for the site to work.
> 
> —
> Reply to this email directly or view it on GitHub.
> 

Sincerely,

Matias Salas

HM: 510-894-1697
MOBILE: 510-688-9996

**Comment by msalas74** [comment]
Yes exactly my point.  And I agree with the cache.  Which brings up my earlier suggestion of incorporating this on the frontend.  If we require the app to reach back to the server for information after a REST call, it would be just a wasted process time which can potentially slow down the response of the app.  We can leverage the users browsers to do the process.


On Aug 9, 2015, at 11:44 AM, rms2219 wrote:

> What Matias said is what I was getting at. There's no reason to store the
> result of the Civic API request when we can just query it on demand using
> the data supplied by our users. We'll store whatever user information is
> necessary to make that query. Another reason you may not want to store this
> data is because we are then responsible for invalidating it once something
> changes at the source. Now, we could certainly cache the result for some
> period so that we can cut down on API requests.
> 
> On Sun, Aug 9, 2015, 2:39 PM Spencer Snygg <notifications@github.com> wrote:
> 
> > Google's CIVIC API would be one stop shopping to get the voter's district
> > divisions. (Federal congressional district, state congressional district,
> > county, city, school district....) Whether we automatically store this
> > information, or have the user type it in, it is critical information. We
> > need to store that information in order for the site to work.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/spencersnygg/VoteWise/issues/13#issuecomment-129224277>
> > .
> >
> —
> Reply to this email directly or view it on GitHub.
> 

Sincerely,

Matias Salas

HM: 510-894-1697
MOBILE: 510-688-9996

**Comment by spencersnygg** [comment]
Please keep in mind that this site is not just for voters to find politicians - it is also for politicians to hear what their voters are saying and do what they want.  If a politician were to try to look at how the people in his district answered questions, and we don't store the users' district information, then we would have to send the address of every single user to the google civic API, then temporarily store all this information, and use that information to return the requested information.  This would have to be done for every politician any time they requested info.  That is a LOT of API requests.  Why is this an improvement over hitting the API once per election cycle per user and storing a very very small amount of information?

**Comment by spencersnygg** [comment]
Maybe I am not explaining myself correctly.  Every ten years, there is a national census.  After the census, areas use the information to redraw district boundaries.  With Google's Civic API, a user can type in their address and google figures out what districts they are in.  This information doesn't change that often.  Most addresses stay in the same districts for a hundred years.  If you are on the border of a district, or if you have politicians doing really weird things to ensure that their party stays in power, then your district could change every ten years.  The district information is what we need.  This is the information that we want to keep.  Google's Civic API also provides information about who is currently in office and sometimes who is running for office.  This information changes every election cycle.  This can be useful, but is not something we need to store, because we should be getting the information from the election officials instead. (however if there was a way of getting the names and contact info of everyone running for office instead of scraping the PDF's from 9000 different election officials, that would be pretty awesome.)

**Comment by msalas74** [comment]
We can tie the district to the question.  If you want to actually have the information of how a specific user answered to a question public, that also can be tied to the question during the polling.  When the question is answered by the user, we can tag that users district info to the question and store it to the database.  Also, if a politician wants to see a specific users district info, that can be obtained dynamically when the politician prompts for the users info.  We would just need the users answer to that specific question since it would be tagged the users district number.  That would not be a lot of API request since it would be event driven from a politician.  There are more voters.

On another note, I'm not sure if voters would like the idea of how they voted on a subject publicly known.

On Aug 9, 2015, at 12:19 PM, Spencer Snygg wrote:

> Please keep in mind that this site is not just for voters to find politicians - it is also for politicians to hear what their voters are saying and do what they want. If a politician were to try to look at how the people in his district answered questions, and we don't store the users' district information, then we would have to send the address of every single user to the google civic API, then temporarily store all this information, and use that information to return the requested information. This would have to be done for every politician any time they requested info. That is a LOT of API requests. Why is this an improvement over hitting the API once per election cycle per user and storing a very very small amount of information?
> 
> —
> Reply to this email directly or view it on GitHub.
> 

Sincerely,

Matias Salas

HM: 510-894-1697
MOBILE: 510-688-9996

**Comment by spencersnygg** [comment]
Users move.  Are you saying that their opinions shouldn't move with them?

I don't understand why storing the user's district info is a problem.  Please explain in more detail.  In your last comment you suggest taking the district info and adding it to every single question.  Why is adding 5 fields for a thousand questions more efficient than adding five fields once?

**Comment by msalas74** [comment]
It depends on how we obtain the information.  If a voter registers with the app, we can collect the users address then.  With the address we can use this single data to collect that users district info from the API.  This can be achieved from the frontend and stored in JSON object.  This info will only be stored in the users local cache.  We can utilize this info when ever the user decides to participate in a question.  When they vote on the question, we can tag a single data that identifies which district they reside in.  This way we are really utilizing one single data that can obtain the rest from the API dynamically.  This can minimize our server requirement and keep the information more up to date.  As for the users moving to another district, this will not address the issue of having some ones vote that does not reside in that district to vote or even ensure that the user will update their information when they vote.  If we require that they input their address each time they vote, we will need to pull the API and then update our database with that information.  That would be on every question for every voter.  This could be costly not just on our bandwidth but server CPU process.  This can equal to lag time.

**Comment by spencersnygg** [comment]
We should only ask for their district once when they are creating their user account.  We have no reason to go back to the API.  The only info we need is the district info which stays the same for a decade.

**Comment by spencersnygg** [comment]
We should prompt them every year to ask if their district has changed and they can choose to check on google, but I see no reason to constantly contact google.

**Comment by msalas74** [comment]
If it's an annual data pull (I'm assuming this would be on the users account registration anniversary date) then I don't see an issue.

**Comment by spencersnygg** [comment]
Cool.

**Comment by spencersnygg** [comment]
And good idea to make it the anniversary date.

**Comment by msalas74** [comment]
thanks.  It was an idea tied to trying to spread that data pull throughout the year instead of having all users pull on the same date.  That would be murder on the server.

**Comment by spencersnygg** [comment]
Upon further research, it seems that Google's Civic API won't do the whole job.  I am looking for other tools to see if there is anything better,

**Comment by spencersnygg** [comment]
I take back my take back:)  The Civic API can do a lot of what we want, but we may have to integrate other tools indirectly and form relationships with other hacker groups to fill out the rest of the info.

**Comment by spencersnygg** [comment]
TO see my discussion with Google about what their API can do, go here:  https://developers.google.com/civic-information/docs/ci_forum

TO see the list of Open Civic Data ID's (OCDID's) that are currently available, go here:  https://github.com/opencivicdata/ocd-division-ids
